core/drivers/crypto/marvell/authenc.c

```diff
@@ -0,0 +1,490 @@
+				     ctx->key_len, ctx->aad_len,
+				     ctx->tag_len, ctx->iv_len,
+				     ctx->iv_data, 0);
+	if (ret != TEE_SUCCESS)
+		goto out_init_err;
+
+	ret = mrvl_ehsm_aes_context_store(&ctx->context_id);
+	if (ret != TEE_SUCCESS)
+		goto out_init_err;
+
+	ctx->is_new = true;
+
+	goto out;
+
+out_init_err:
+	free(ctx->key);
+out_key_err:
+	free(ctx->aad_data);
+out_aad_err:
+	free(ctx->iv_data);
+out:
+	mrvl_ehsm_cryp_unlock();
+	return ret;
+}
+
+static TEE_Result
+mrvl_ae_update_aad(struct drvcrypt_authenc_update_aad *dupdate)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct mrvl_ae_ctx *c = to_mrvl_ae_ctx(dupdate->ctx);
+	struct mrvl_cryp_context *ctx = &c->cryp_ctx;
+
+	if ((dupdate->aad.length + ctx->aad_cur_idx) <= ctx->aad_len) {
+		memcpy(ctx->aad_data + ctx->aad_cur_idx, dupdate->aad.data,
+		       dupdate->aad.length);
+		ctx->aad_cur_idx += dupdate->aad.length;
+	} else {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return ret;
+}
+
+static TEE_Result
+mrvl_ae_update_payload(struct drvcrypt_authenc_update_payload *dupdate)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct mrvl_ae_ctx *c = to_mrvl_ae_ctx(dupdate->ctx);
+	struct mrvl_cryp_context *ctx = &c->cryp_ctx;
+	uint8_t *payload_buf = NULL;
+	uint8_t *encdata_buf = NULL;
+	size_t payload_buf_len = 0;
+	size_t encdata_buf_len = 0;
+	uint32_t context_id = 0;
+	size_t iv = 0;
+
+#if defined(PLATFORM_FLAVOR_cn10ka) || defined(PLATFORM_FLAVOR_cn10kb) || \
+	defined(PLATFORM_FLAVOR_cnf10ka) || defined(PLATFORM_FLAVOR_cnf10kb)
+	return TEE_ERROR_NOT_IMPLEMENTED;
+#endif
+	mrvl_ehsm_cryp_lock();
+
+	ret = mrvl_ehsm_aes_context_load(ctx->context_id);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	if (ctx->is_new) {
+		/* IV is not 12-byte, load IV part of AES operation payload */
+		if (ctx->iv_len != 12)
+			iv = ctx->iv_len;
+		else
+			iv = 0;
+
+		payload_buf_len = iv + ctx->aad_len + dupdate->src.length;
+
+		payload_buf = mem_alloc(payload_buf_len);
+		if (!payload_buf) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out;
+		}
+
+		encdata_buf_len = iv + ctx->aad_len + dupdate->dst.length;
+
+		encdata_buf = mem_alloc(encdata_buf_len);
+		if (!encdata_buf) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out_enc_buf_err;
+		}
+
+		if (ctx->iv_len != 12)
+			memcpy(payload_buf, ctx->iv_data, iv);
+
+		memcpy(payload_buf + iv, ctx->aad_data, ctx->aad_cur_idx);
+		memcpy(payload_buf + iv + ctx->aad_cur_idx, dupdate->src.data,
+		       dupdate->src.length);
+
+		ret = mrvl_ehsm_aes_gcm_update_payload(payload_buf,
+						       dupdate->src.length,
+						       encdata_buf,
+						       dupdate->dst.length,
+						       ctx->is_new);
+
+		if (ret == TEE_SUCCESS)
+			memcpy(dupdate->dst.data, encdata_buf + iv +
+			       ctx->aad_cur_idx, dupdate->dst.length);
+
+		ctx->is_new = false;
+
+		context_id = ctx->context_id;
+		ret = mrvl_ehsm_aes_context_store(&ctx->context_id);
+		if (ret != TEE_SUCCESS)
+			goto out_ctx_err;
+
+		ret = mrvl_ehsm_aes_context_release(context_id);
```

Comment:
Should `mrvl_ehsm_aes_context_release()` be also called for the other error cases?

Reply:
For update payload error case mrvl_ehsm_aes_context_release not be called, as context will be lost inside AES engine.